[Evo] Ref #5274 - Request access to each metadata

nabdali opened the pull request:
- Add list of field in data access request object
- Implement DCModel class to get all necessary value and clean keys
- Add fields on create and update operation of data access request object
- Display a field list of model with checkbox in data request form

nabdali committed:
[Evo] Ref #5274 - Request access to each metadata

- Add list of field in data access request object
- Implement DCModel class to get all necessary value and clean keys
- Add fields on create and update operation of data access request object
- Display a field list of model with checkbox in data request form

nabdali commented on src/main/kotlin/org/ozwillo/dcimporter/model/datacore/DCModel.kt:
```diff
@@ -0,0 +1,62 @@
+import java.util.Objects
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+class DCModel : Comparable<DCModel> {
```
This is a copy of DCModel class from ozwillo spring integration lib,  I added get/set JsonProperty annotations to have cleans keys (without :) of this object in front.

bobeal replied:
why not a data class ?

nabdali replied:
I have tried but i had a parsing error, similar of this error in the link https://github.com/FasterXML/jackson-databind/issues/1122

nabdali commented on src/main/kotlin/org/ozwillo/dcimporter/service/DatacoreService.kt:
```diff
@@ -255,13 +255,39 @@ class DatacoreService(private val kernelProperties: KernelProperties) {
         }
     }
 
+    fun findModel(project: String, model: String): Flux<DCModel>{
```
Not used yet, but we can use it later !

nabdali commented on src/main/kotlin/org/ozwillo/dcimporter/service/DatacoreService.kt:
```diff
@@ -255,13 +255,39 @@ class DatacoreService(private val kernelProperties: KernelProperties) {
         }
     }
 
+    fun findModel(project: String, model: String): Flux<DCModel>{
```
```suggestion
    fun findModel(project: String, model: String): Mono<DCModel>{
```

bobeal replied:
yes, Mono, not Flux

nabdali commented on src/main/resources/public/src/views/DataRequest.vue:
```diff
@@ -118,13 +130,21 @@
         },
         methods: {
+            handleModelSelected(model) {
+                model.fields.map(field => Vue.util.defineReactive(field, 'requested' , true))
```
defineReactive is a function in a tools of vueJS for adding a propety to an vueJS data object

laurianemas commented on src/main/kotlin/org/ozwillo/dcimporter/model/datacore/DCModel.kt:
```diff
@@ -0,0 +1,62 @@
+        @set:JsonProperty("dcmf:name")
+        var name: String? = null
+    }
+}
```
Why not. Here, we simplify keys in front but we add a new class model instead of using DCBusinessResourceLight. That means that DC Models are treated differently from other DC objects and loose the logic of unique DC object type as DCBusinessResourceLight. Does it worth it ? Is the complexes json keys with ":" a problem in front ?

bobeal replied:
good point

laurianemas commented on src/main/kotlin/org/ozwillo/dcimporter/web/Routes.kt:
```diff
@@ -75,6 +74,7 @@ class Routes(
             PUT("/type/{type}", datacoreHandler::updateResourceWithOrganization)
             GET("/organizations", datacoreHandler::getAllOrganization)
             GET("/models", datacoreHandler::getModels)
+            GET("/model/{project}/{type}", datacoreHandler::getModel)
```
Is project really needed in the URI ? For others handlers the project value is define in back. And I think we can use the mother project oasis.main to deal with all DC models.

laurianemas commented on src/main/resources/public/src/views/DataRequest.vue:
```diff
@@ -84,7 +96,7 @@
             disabled: function(){
-                return (this.organizationSelected == false || this.dataAccessRequest.email == '' || this.modelSelected == '')
+                return (this.organizationSelected === false || this.dataAccessRequest.email === '' || this.modelSelected === '')
             }
```
If you initialize modelSelected to null value, you need to add it in the computed disabled condition otherwise users will be able to submit forms without selecting models on first screen opening.

laurianemas commented on src/main/resources/public/src/views/DataRequest.vue:
```diff
@@ -31,17 +31,29 @@
                     :data="models"
-                    @hit="modelSelected = $event"/>
+                    :serializer="displayingResultOfModelSearch"
+                    @hit="handleModelSelected($event)"/>
```
Find a little problem here. If we select a model, this one is saved in modelSelected and the "Fields" fields are displayed. But v-on:hit do not handle deletion in the input. In consequences if a user select a model and then delete it, "Fields" fields are still displayed and modelSelected still fullfill with previous selection, that's not very clean. Plus, as the submit disabled property is based on modelSelected, users can submit without finalize their model selection (the previous choice will be the one saved in db).

You can use v-on-input event to watch change in the input and change modelSelected value to null when modelSearch is empty (or directly use the watcher I suppose but I always have problems to deal with the debounce system).

nabdali replied:
I did this in the get model method that is called on change of modelSelected value

nabdali committed:
[Evo] Fix #5274 - Find a DC model by type only

- Fix verification of selected model in data access request form
-  Advocate in review of PR #14 to put a static model project in back to finding a model
